go: build PIE by default

LunNova opened the pull request:
In #439314 we configured `gcc` to enable PIE by default, without needing a wrapper's involvement.  
This change aims to achieve the same outcome for go.

This change was originally going to be included in #442510 which drops the PIE hardening flag as obsolete, but has been split since it was getting a bit too complicated to roll together.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-linux -> aarch64-multiplatform cross
  - [x] x86_64-linux -> ppc64-elfv1 cross
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [x] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

LunNova committed:
go: teach internal linker to use correct ELF interpreter

Teaches go's internal linker to use the GO_LDSO env var as the path for the dynamic linker, instead of defaulting to an FHS /lib path.
GO_LDSO is automatically set to the appropriate value by checking $NIX_CC/nix-support/dynamic-linker
External linking is set as the default for cross compile situations
where CGO is supported.
Teaching go to correctly handle cross
in the internal linker is hard, and our system linker already knows
how to find the right ELF interpreter.
For cross situations where CGO is not supported we assume either non-ELF
binaries or static binaries will be produced.

LunNova committed:
go: position independent executables by default

Followup to setting --enable-default-pie for our gcc,
we teach go to build PIE by default when supported.

LunNova committed:
linkerd: remove unnecessary buildmode=pie

This is no longer needed now that our go produces PIE by default

LunNova commented:
I have all of `nix build .#{,pkgsCross.aarch64-multiplatform,pkgsMusl.,pkgsStatic.}{syncthing,clickhouse-backup,go}` building with the latest commit, not sure if the approach is ok.

LunNova commented on pkgs/development/compilers/go/1.25.nix:
```diff
@@ -104,13 +111,20 @@ stdenv.mkDerivation (finalAttrs: {
       export CC=${buildPackages.stdenv.cc}/bin/cc
+      # Prefer external linker for cross, teaching go's internal linker to do cross right is hard
+      # This flag is baked into the compiler and not just for bootstrap
+      export GO_EXTLINK_ENABLED=1
```
Should be fixed now, I was able to `nix build .#pkgsCross.ppc64-elfv1.{buildPackages.go,clickhouse-backup}`, as well as the previous testcases.

OPNA2608 reviewed (approved):
I don't really feel qualified to comment on the Go patching, but I know that you and emily (IIRC?) were discussing the approach for this in the staging room awhile ago, so I *assume* it's fine. Doesn't seem to break POWER cross at least. :)

philiptaron reviewed (approved):
Doing a build on `nixosTests.teleport` (uses Go 1.24) and `tailscale` (uses Go 1.25) and `nixosTests.traefik` (uses some combo?!)

philiptaron commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 442965 --package tailscale --package nixosTests.traefik`
Commit: `87175fde5623f82474cfe70a9572c5669eb2f41e`

---
### `x86_64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>tailscale</li>
    <li>tailscale.derper (tailscale.derper.derper)</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 test built:</summary>
  <ul>
    <li>nixosTests.traefik</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-linux`</summary>
<details>
<summary>tailscale</summary>
<pre>ok  	tailscale.com/tempfork/gliderlabs/ssh	0.003s
ok  	tailscale.com/tempfork/heap	0.004s
ok  	tailscale.com/tempfork/sshtest/ssh	0.430s
ok  	tailscale.com/tka	0.527s
ok  	tailscale.com/tsconsensus	0.013s
ok  	tailscale.com/tsnet	0.007s [no tests to run]
ok  	tailscale.com/tstest	0.016s
--- FAIL: TestInQemu (0.00s)
    --- FAIL: TestInQemu/386 (0.02s)
        qemu_test.go:65: failed: # tailscale.com/tstest/archtest
            default PIE binary requires external (cgo) linking, but cgo is not enabled
            FAIL	tailscale.com/tstest/archtest [setup failed]
            FAIL
        qemu_test.go:69: output didn&#x27;t contain &quot;I am linux/386&quot;: # tailscale.com/tstest/archtest
            default PIE binary requires external (cgo) linking, but cgo is not enabled
            FAIL	tailscale.com/tstest/archtest [setup failed]
            FAIL
FAIL
FAIL	tailscale.com/tstest/archtest	0.024s
FAIL</pre>
</details>
</details>

LunNova commented:
Tailscale is doing some interesting cross-compiling internally and qemu testing that this has failed to handle because it has nothing to do with our selected go compiler's targetPlatform.

https://github.com/tailscale/tailscale/blob/ad6cf2f8f369ae54652a0808cda872ca558ab429/tstest/archtest/qemu_test.go#L19-L72

philiptaron commented:
What's the right thing to do? Can we push a patch to disable that test?

LunNova commented:
> What's the right thing to do? Can we push a patch to disable that test?

I having the enable-default-pie patch not apply when cross is being done in a not-nixpkgs-stdenv-aware way would solve this more generally, experimenting.

LunNova committed:
go: only enable PIE by default when building for GOARCH matching target

We rely on teaching go to use the correct ELF interpreter
for PIE in a way that can't work for packages that 
go off and do their own GOARCH=something cross builds
without touching nixpkgs cross machinery,
so update the default pie patch to bake targetPlatform's
GOARCH and only activate for it.

tailscale works as a testcase for this as it cross compiles
many arches in tstest/archtest/qemu_test.go

LunNova commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 442965 --package clickhouse-backup --package pkgsMusl.syncthing --package pkgsCross.aarch64-multiplatform.syncthing --package pkgsCross.aarch64-multiplatform.go --package pkgsCross.ppc64-elfv1.buildPackages.go --package pkgsStatic.syncthing --package pkgsCross.aarch64-multiplatform.clickhouse-backup --package go --package pkgsMusl.go --package nixosTests.traefik --package pkgsStatic.clickhouse-backup --package pkgsMusl.clickhouse-backup --package pkgsCross.ppc64-elfv1.clickhouse-backup --package go.passthru.tests.clickhouse-backup-is-pie --package tailscale --package syncthing --package pkgsStatic.go`
Commit: `dbb57d8d86f6fcc92d28c3d6835a83d85bb5e4b5`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 test built:</summary>
  <ul>
    <li>nixosTests.traefik</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 17 packages built:</summary>
  <ul>
    <li>clickhouse-backup</li>
    <li>go</li>
    <li>go.passthru.tests.clickhouse-backup-is-pie</li>
    <li>pkgsCross.aarch64-multiplatform.clickhouse-backup</li>
    <li>pkgsCross.aarch64-multiplatform.go</li>
    <li>pkgsCross.aarch64-multiplatform.syncthing</li>
    <li>pkgsCross.ppc64-elfv1.buildPackages.go</li>
    <li>pkgsCross.ppc64-elfv1.clickhouse-backup</li>
    <li>pkgsMusl.clickhouse-backup</li>
    <li>pkgsMusl.go</li>
    <li>pkgsMusl.syncthing</li>
    <li>pkgsStatic.clickhouse-backup</li>
    <li>pkgsStatic.go</li>
    <li>pkgsStatic.syncthing</li>
    <li>syncthing</li>
    <li>tailscale</li>
    <li>tailscale.derper (tailscale.derper.derper)</li>
  </ul>
</details>

philiptaron reviewed (approved):
Neat.
